### PR TITLE
Added script to validate .mo/.mos file 

### LINKED ISSUE #145: Model validation
This issue is to extend BuildingsPy to check/validate mos and mo files so they comply with JModelica

### PR DESCRIPTION
**Pull request by tsnouidui**
This closes #145.

### CODE REVIEW

File: buildingspy/development/check_parameters.py [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+                datefmt='%m/%d/%Y %I:%M:%S %p')
+stderrLogger = log.StreamHandler()
+stderrLogger.setFormatter(log.Formatter(log.BASIC_FORMAT))
+log.getLogger().addHandler(stderrLogger)
+
+import os
+import re
+import sys
+
+class Validator(object):
```
Comment (mwetter) [inline buildingspy/development/check_parameters.py]: Shouldn't this code be in `buildingspy/development/validator.py`, as this file already has a `Validator` class.

File: buildingspy/development/check_parameters.py [hunk 2]
```diff
@@ -0,0 +1,476 @@
...
+# TSNouidui@lbl.gov                            2017-01-24
+###########################################################
+#
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+import logging as log
+
+log.basicConfig(filename='check_parameters.log', filemode='w',
```
Comment (mwetter) [inline buildingspy/development/check_parameters.py]: If you need a log file, please use `buildingspy/io/reporter`. This way, if we want to use a logger for BuildingsPy, we can refactor one class and don't have different implementations in the same library.

File: buildingspy/development/check_parameters.py [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+        :param name: Parameter name.
+        :param value: Value found in mos file.
+        :param modelPath: Path to mo file.
+        :param mos_file: Path to mos file.
+    
+         """
+        
+        log.error("Found mo_file: {!s} without **experiment** annotation **{!s}**.".format(modelPath, self._capitalize_first(name)))
+        log.error("The parameter **"+name+"** with value : {!s} is however defined in the .mos file {!s}.".format(value, mos_file)) 
+        sys.exit(1)
```
Comment (mwetter) [inline buildingspy/development/check_parameters.py]: Raise an exception, but never call `sys.exit(1)` as raising exceptions allows the error to be handled.

File: buildingspy/development/check_parameters.py [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+                            self._check_experiment(name, val, value, modelPath, mos_file)
+                            found = True
+                    
+                elif value == ""+name+"":
+                    self._wrong_literal(modelPath, name)
+    
+                
+            f.close()
+    
+    def validateModelParameters(self, rootDir):
```
Comment (mwetter) [inline buildingspy/development/check_parameters.py]: Python docstring missing

File: buildingspy/tests/test_development_check_parameters.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+       :mod:`buildingspy.development.check_parameters.
+    """
+
+    def test_validateModelParameters(self):
+        import os
+        import buildingspy.development.check_parameters as v
+        val = v.Validator()
+        myMoLib = os.path.join("buildingspy", "tests", "MyModelicaLibrary")
+        val.validateModelParameters(myMoLib)
+
```
Comment (mwetter) [inline buildingspy/tests/test_development_check_parameters.py]: Add also more tests, for example, add tests that verify that you catch wrong entries. The test you have now only shows that a correct formatted library passes, but it does not verify that a wrong entry is catch. (You may need to refactor your code a bit to test whether wrong entries are caught.)

File: buildingspy/development/check_parameters.py (hunk 2 above)
Comment (tsnouidui) [inline buildingspy/development/check_parameters.py]: OK

File: buildingspy/development/check_parameters.py (hunk 1 above)
Comment (tsnouidui) [inline buildingspy/development/check_parameters.py]: Makes sense. I will see if I can use this class instead.

File: buildingspy/development/check_parameters.py (hunk 3 above)
Comment (tsnouidui) [inline buildingspy/development/check_parameters.py]: Good idea.

File: buildingspy/development/check_parameters.py (hunk 4 above)
Comment (tsnouidui) [inline buildingspy/development/check_parameters.py]: OK

File: buildingspy/tests/test_development_check_parameters.py (hunk 5 above)
Comment (tsnouidui) [inline buildingspy/tests/test_development_check_parameters.py]: Will see how to approach this best. Thanks

**Comment by thorade** [comment]
Thierry, I have assigned PR #147 to you, suggesting to add the same header to your files as was added in the master, too.

File: buildingspy/tests/test_development_Validator.py [hunk 6]
```diff
@@ -9,19 +9,37 @@
 from future import standard_library
 standard_library.install_aliases()
 from builtins import *
-from io import open
 # end of from future import
+from jinja2 import Template
```
Comment (thorade) [inline buildingspy/tests/test_development_Validator.py]: this adds jinja2 as a dependency

**Comment by tsnouidui** [comment]
@mwetter I addressed your comments. Could you please review?

**Comment by thorade** [comment]
One more comment: You have a function capitalize_first:
https://github.com/lbl-srg/BuildingsPy/pull/146/files#diff-ab7c00b10f083eaa6adc6839f2cf9023R245

Would one of the following functions from the Python standard library work for you instead?
https://docs.python.org/3/library/stdtypes.html#str.capitalize
https://docs.python.org/3/library/stdtypes.html#str.title

**Comment by tsnouidui** [comment]
@thorade Thanks for the two functions. The problem is that the functions will capitalize the first letter but make the other letters lowercase e.g. `stopTime` will in this case become `Stoptime`. This is not what I want. I want `stopTime` to be `StopTime`. Are you aware of any function which could achieve this?

**Comment by thorade** [comment]
I think then your approach is the way to go

File: buildingspy/development/validator.py [hunk 7]
```diff
@@ -177,3 +179,377 @@ def _validateHTML(self, moFile):
...
+                    self._check_tolerance(content, name, value, mos_file)     
+                else:
+                    found = False
+                    while found == False and i<len(content):
+                        line = content[i]
+                        i += 1
+                        # Remove white spaces
+                        line.replace(" ", "")
+                        
+                        if ""+name+"="+"" in line.replace(" ", ""):
```
Comment (mwetter) [inline buildingspy/development/validator.py]: Do you add zero characters to a string here?

File: buildingspy/development/validator.py [hunk 8]
```diff
@@ -177,3 +179,377 @@ def _validateHTML(self, moFile):
...
+                            found = True
+                            pTime    = re.compile(r"[\d\S\s.,]*("+name+"=)([\d]*[.]*[\d]*[eE]*[+|-]*[\d]*[*]*[\d]*[.]*[\d]*[eE]*[+|-]*[\d]*)[\S\s.,]*")
+                            mTime    = pTime.match(line)
+                            value = mTime.group(2)
+                            self._check_tolerance(content, name, value, mos_file)  
+                            #startTime = startTime[:-1]
+                        if ""+name+"="+name+"" in line.replace(" ", ""):
```
Comment (mwetter) [inline buildingspy/development/validator.py]: Zero characters added to string?

File: buildingspy/development/validator.py [hunk 9]
```diff
@@ -177,3 +179,377 @@ def _validateHTML(self, moFile):
...
+                    found = False
+                    while found == False and i<len(content):
+                        line = content[i]
+                        i += 1
+                        # Remove white spaces
+                        line.replace(" ", "")
+                        
+                        if ""+name+"="+"" in line.replace(" ", ""):
+                            found = True
+                            pTime    = re.compile(r"[\d\S\s.,]*("+name+"=)([\d]*[.]*[\d]*[eE]*[+|-]*[\d]*[*]*[\d]*[.]*[\d]*[eE]*[+|-]*[\d]*)[\S\s.,]*")
```
Comment (mwetter) [inline buildingspy/development/validator.py]: Is this `pTime` reg expression the same as on line 430? It seems a bit different, but I don't see why and what it is doing as there is no comment and otherwise the code seems to be the same as around line 430. Please correct to be clearer (and give others a chance to edit your code if needed).

File: buildingspy/development/validator.py [hunk 10]
```diff
@@ -177,3 +179,377 @@ def _validateHTML(self, moFile):
...
+                        elif(name=="tolerance"):
+                            value="0.0"
+                            print("Should be coming here")
+                            self._wrong_parameter (mos_file, name, value)
+    
+            except AttributeError:
+                value = "NA"
+                N_mos_defect += 1
+                
+            if value != "NA" and value != ""+name+"":
```
Comment (mwetter) [inline buildingspy/development/validator.py]: Why `""+name"+""`, and why NA? Is this a string variable that is used to test something? If so, please use `None` or a boolean flag.

File: setup.py [hunk 11]
```diff
@@ -17,8 +17,12 @@ def read(fname):
     keywords = "modelica dymola openmodelica mat",
     url = "http://simulationresearch.lbl.gov/modelica/",
     install_requires = ['future',
+                        'gitpython',
+                        'jinja2',
+                        'matplotlib',
+                        'numpy',
                         'pytidylib',
-                        'gitpython'],
+                        'scipy'],
```
Comment (mwetter) [inline setup.py]: What of your code requires `scipy`? This is a big dependencies to check the value of a tolerance.

File: buildingspy/development/validator.py [hunk 12]
```diff
@@ -177,3 +179,377 @@ def _validateHTML(self, moFile):
...
+    
+         """
+         
+        if ("" + name + "=" + "" == "tolerance="):
+            if(float(value) > 1e-6):
+                s = ("Found mos file=" + str(mos_file) + " with tolerance=" + str(value) + ".\n"
+                    "The tolerance found is bigger than 1e-6, the maximum required by "
+                    "JModelica for unit tests.\n")
+                raise ValueError(s)
+            elif(float(value) == 0.0):
```
Comment (mwetter) [inline buildingspy/development/validator.py]: Don't check floats for equality. Use `None` or a flag if it is not assigned.

File: buildingspy/development/validator.py [hunk 13]
```diff
@@ -177,3 +179,377 @@ def _validateHTML(self, moFile):
...
+                s = ("Found mos file=" + str(mos_file) + " without tolerance specified.\n" + 
+                    "A maximum tolerance of 1e-6 is required by JModelica for unit tests.\n")
+                raise ValueError(s)
+        if ("" + name + "=" + "" == "stopTime="):
+            if(float(value) == 0.0):
```
Comment (mwetter) [inline buildingspy/development/validator.py]: Don't use `float` for equality.

File: buildingspy/development/validator.py [hunk 14]
```diff
@@ -177,3 +179,377 @@ def _validateHTML(self, moFile):
...
+        :param model_path: Path to mo file.
+        :param mos_file: Path to mos file.
+    
+         """
+        
+        s = ("Found mo file=" + str(model_path) + " without parameter " 
+             + self._capitalize_first(name) + " defined.\n" 
+            + "The parameter name " + name + " is defined in the mos file=" 
+            + str(mos_file) + " and hence must be defined in the mo file.\n")
+        raise ValueError(s)
```
Comment (mwetter) [inline buildingspy/development/validator.py]: Why do you need `str()` for some but not for other arguments? Maybe using `...format(...)` would be easier (and more standard than adding strings).

File: buildingspy/development/validator.py [hunk 15]
```diff
@@ -177,3 +179,377 @@ def _validateHTML(self, moFile):
...
+        
+        if("*" in str(val)):
+            s = ("Found mo file=" + str(model_path) + " with experiment annotation " + self._capitalize_first(name) + ".\n" 
+                + self._capitalize_first(name) + " contains invalid expressions such as x * y. Only literal expressions are allowed " 
+                + "by JModelica and OpenModelica unit tests.\n")
+            raise ValueError(s) 
+    
+        delta = abs(eval(val) - eval(value))
+        
+        if (delta!=0):
```
Comment (mwetter) [inline buildingspy/development/validator.py]: Don't check floats for equality.

File: .travis.yml [hunk 16]
```diff
@@ -38,11 +38,11 @@ before_install:
 
 install:
   - pip install --upgrade pip setuptools wheel
-  - pip install --only-binary=numpy,scipy,matplotlib numpy scipy matplotlib
-  - pip install future gitpython pytidylib
+  - pip install --only-binary=matplotlib,numpy,scipy matplotlib numpy scipy
```
Comment (mwetter) [inline .travis.yml]: Why is this needed for your issue? If it is not related to your issue, please make such changes on a new issue and branch.

File: .travis.yml [hunk 17]
```diff
@@ -38,11 +38,11 @@ before_install:
...
   - pip install .
 
 # Execute tests
 script:
   - omc --version
-  - make doctest unittest
+  - make doctest
+  - make unittest
```
Comment (mwetter) [inline .travis.yml]: Any reason for this change in this issue?

**Comment by thorade** [comment]
@mwetter Sorry, all changes in travis.yml and setup.py were done by me. I wanted to add the future imports before the new files are on the master and then I also added jinja2 to make travis pass and then I noticed that dependencies mentioned in setup.py do not have to be mentioned in travis.yml and then I sorted them alphabetically and then ...

All changes in setup.py and travis.yml can be reverted, except adding jinja2 as a dependency.

**Comment by thorade** [comment]
The conflicts can probably be resolved by running autopep as done on master:
`make pep8 PEP8_CORRECT_CODE=true`

File: buildingspy/development/validator.py (hunk 15 above)
Comment (tsnouidui) [inline buildingspy/development/validator.py]: OK.

File: buildingspy/development/validator.py (hunk 14 above)
Comment (tsnouidui) [inline buildingspy/development/validator.py]: Using format now.

File: buildingspy/development/validator.py (hunk 12 above)
Comment (tsnouidui) [inline buildingspy/development/validator.py]: OK.

File: buildingspy/development/validator.py (hunk 13 above)
Comment (tsnouidui) [inline buildingspy/development/validator.py]: OK.

File: buildingspy/development/validator.py (hunk 7 above)
Comment (tsnouidui) [inline buildingspy/development/validator.py]: Shoudn't have been there. Removed.

File: buildingspy/development/validator.py (hunk 9 above)
Comment (tsnouidui) [inline buildingspy/development/validator.py]: I simplified the implementation and got rid of the regular expression.

File: buildingspy/development/validator.py (hunk 8 above)
Comment (tsnouidui) [inline buildingspy/development/validator.py]: Shoudn't have been there. Removed.

File: buildingspy/development/validator.py (hunk 10 above)
Comment (tsnouidui) [inline buildingspy/development/validator.py]: Done.

**Comment by tsnouidui** [comment]
@mwetter: I revised the implementation and made the desired changes.

File: .travis.yml (hunk 16 above)
Comment (tsnouidui) [inline .travis.yml]: Reverted this.

File: .travis.yml (hunk 17 above)
Comment (tsnouidui) [inline .travis.yml]: Reverted this.

File: setup.py (hunk 11 above)
Comment (tsnouidui) [inline setup.py]: Reverted this.

**Comment by tsnouidui** [comment]
@mwetter This is ready to be reviewed. Now the code also checks the existence of a `StopTime` in the .mo file and rejects the model if not present.
